README.md
# ThorPerformance

ThorPerformance is a DataBroker addon that shows performance statistics.


CHANGELOG.md
# ThorPerformance changelog

## 1.0.0 (2022-11-25)

- First release.


ThorPerf.lua
-----------------------------------
-- Setting up scope, upvalues and libs
-----------------------------------

local AddonName, ThorPerf = ...;
LibStub("AceEvent-3.0"):Embed(ThorPerf);
LibStub("AceTimer-3.0"):Embed(ThorPerf);

local LibCrayon = LibStub("LibCrayon-3.0");

local _G = _G; -- I always use _G.FUNC when I call a Global. Upvalueing done here.
local format = string.format;

-------------------------------
-- Registering with iLib
-------------------------------

LibStub("iLib"):Register(AddonName, nil, ThorPerf);

-----------------------------------------
-- Variables, functions and colors
-----------------------------------------

local hugeFramerate = 48; -- at 24fps, the fps becomes yellow, higher values color it green, lower values red
local hugeLatency = 200; -- at 100ms, the latency becomes yellow, lower than 100ms colors it green, higher values red

local UpdateTimer;
local Mods = {};

local ThorPerfFramerate = 0;
local ThorPerfLatencyHome = 0;
local ThorPerfLatencyWorld = 0;
local ThorPerfUpload = 0;
local ThorPerfDownload = 0;

local COLOR_GOLD = "|cfffed100%s|r";

------------------------------
-- Formatting Functions
------------------------------

local function format_kbs(value)
	return ("%.1f "..COLOR_GOLD):format(value, "kb/s");
end

local function format_latency(value, space)
	return ("|cff%s%d|r"..(space and "" or " ")..COLOR_GOLD):format(LibCrayon:GetThresholdHexColor(hugeLatency - value, hugeLatency), value, "ms");
end

local function format_fps(value, space)
	return ("|cff%s%.1f|r"..(space and "" or " ")..COLOR_GOLD):format(LibCrayon:GetThresholdHexColor(value, hugeFramerate), value, "fps");
end

-----------------------------
-- Setting up the LDB
-----------------------------

ThorPerf.ldb = LibStub("LibDataBroker-1.1"):NewDataObject(AddonName, {
	type = "data source",
	text = "",
});

ThorPerf.ldb.OnClick = function(_, button)
end

ThorPerf.ldb.OnEnter = function(anchor)
	if( ThorPerf:IsTooltip("Main") ) then
		return;
	end
	ThorPerf:HideAllTooltips();
	
	local tip = ThorPerf:GetTooltip("Main", "UpdateTooltip");
	tip:SmartAnchorTo(anchor);
	tip:SetAutoHideDelay(0.25, anchor);
	
	tip:Show();
end

ThorPerf.ldb.OnLeave = function() end -- some display addons refuse to display brokers when this is not defined

function ThorPerf:Boot()
	self:RefreshTimer();
	self:UnregisterEvent("PLAYER_ENTERING_WORLD");
end
ThorPerf:RegisterEvent("PLAYER_ENTERING_WORLD", "Boot");

function ThorPerf:RefreshTimer(timeToPass)
	if( UpdateTimer ) then
		self:CancelTimer(UpdateTimer);
		UpdateTimer = nil;
	end
	
	if( not timeToPass ) then
		timeToPass = 2.5;
	end
	
	UpdateTimer = self:ScheduleRepeatingTimer("UpdateData", timeToPass);
	self:UpdateData();
end

function ThorPerf:UpdateBroker()
	self.ldb.text =
		(format_fps(ThorPerfFramerate, true).." " or "")..
		(format_latency(ThorPerfLatencyHome, true).." " or "")..
		(format_latency(ThorPerfLatencyWorld, true).." " or "");
end

function ThorPerf:UpdateData()	
	ThorPerfFramerate = _G.GetFramerate();
	ThorPerfDownload, ThorPerfUpload, ThorPerfLatencyHome, ThorPerfLatencyWorld = _G.GetNetStats();
	
	self:CheckTooltips("Main");
	self:UpdateBroker();
end

function ThorPerf:UpdateTooltip(tip)
	local firstDisplay = tip:GetLineCount() == 0;
	if( firstDisplay ) then
		tip:SetColumnLayout(2, "LEFT", "RIGHT");

		tip:AddLine("ThorPerf");
		
		tip:AddLine();
		tip:AddLine();
		tip:AddLine();
		tip:AddLine();
		tip:AddLine();
		tip:AddLine();

		tip:SetCell(3, 1, (COLOR_GOLD):format("Framerate"), nil, "LEFT");
		tip:SetCell(4, 1, (COLOR_GOLD):format("Latency Home"), nil, "LEFT");
		tip:SetCell(5, 1, (COLOR_GOLD):format("Latency World"), nil, "LEFT");
		tip:SetCell(7, 1, (COLOR_GOLD):format("Stream download"), nil, "LEFT");
		tip:SetCell(6, 1, (COLOR_GOLD):format("Stream upload"), nil, "LEFT");
	end

	tip:SetCell(3, 2, format_fps(ThorPerfFramerate), nil, "RIGHT");
	tip:SetCell(4, 2, format_latency(ThorPerfLatencyHome), nil, "RIGHT");
	tip:SetCell(5, 2, format_latency(ThorPerfLatencyWorld), nil, "RIGHT");
	tip:SetCell(6, 2, format_kbs(ThorPerfDownload), nil, "RIGHT");
	tip:SetCell(7, 2, format_kbs(ThorPerfUpload), nil, "RIGHT");
end


Libs/iLib/iLib.lua
local MAJOR_VERSION, MINOR_VERSION = "iLib", 600100
if( not LibStub ) then error(MAJOR_VERSION.." requires LibStub"); end

local iLib, oldLib = LibStub:NewLibrary(MAJOR_VERSION, MINOR_VERSION);
if( not iLib ) then return; end

local AceComm = LibStub("AceComm-3.0");

local _G = _G;

-------------------
-- Variables
-------------------

local ME_UPDATE, EQUAL, USER_UPDATE = 1, 2, 3; -- currently we only use USER_UPDATE to check if we need to send an update message

local mods_changed = false; -- determines if new mods registered with the iLib
local player = _G.GetUnitName("player");

local inGuild = false; -- determines if we are in a guild
local inGroup = false; -- determines if we are in a group
local inInstanceGroup = false; -- determines if we are in an instance group
local inInstanceType = "none"; -- determines our instance type

local Embed; -- will become a function later

--------------------------
-- Lib initializing
--------------------------

-- hide this out from iLib table
local mt_index = {
	mods = iLib.mods or {},
	warn = iLib.warn or {}, -- warn list for addon update messages
	touch = iLib.touch or {}, -- stores whisper list, username/bool pairs
	update = iLib.update or {},
};
setmetatable(iLib, {__index = mt_index});

if( not getmetatable(iLib.mods) ) then
	setmetatable(iLib.mods, {__newindex = function(t, k, v) -- new indexes in iLib.mods will cause addonsChanged to be true
		rawset(t, k, v);
		mods_changed = true;
	end})
end

------------------------
-- Addon messages
------------------------

-- iLib addon messages
-- ask - asks other players if we are up to date
--  ?%addon1%version1%addon2%version2%..%addonN%versionN
-- update - the respond only(!) includes addons which can be updated
--   If playerX asked for updates, playerY will respond. If playerY isn't up to date, too, there is another !-message fired by playerZ
--  !%addon1%version1%..%addonN%versionN
-- alist - requests the list of addons registered with iLib
--  ;
-- addons - answers with the addon list
--  :%addon1%addon2%..%addonN
-- touch - requests if a whisper or mouseover unit has iLib
--  t
-- touch resp - answer if iLib is present
--  y

local _encode, _decode;
local send_msg_ask, send_msg_alist, send_msg_update, send_msg_touch, send_msg_touch_resp;
do
	-- stores our messages
	local ask_message;
	local addon_message;
	
	-- creates the message strings if our addons have changed
	local function compile_addon_string()
		if( not mods_changed ) then
			return;
		end
		
		ask_message = "?";
		addon_message = ":";
		
		for addon, version in pairs(iLib.mods) do
			ask_message = ask_message.."%"..addon.."%"..version;
			addon_message = addon_message.."%"..addon;
		end
		
		mods_changed = true;
	end
	
	-- sends our ask message
	send_msg_ask = function(chat, user)
		compile_addon_string();
		AceComm.SendCommMessage(iLib, "iLib", _encode(ask_message), chat, (chat == "WHISPER" and user or nil), "BULK");
	end
	
	-- sends our addon message
	send_msg_alist = function(chat, user)
		compile_addon_string();
		AceComm.SendCommMessage(iLib, "iLib", _encode(addon_message), chat, (chat == "WHISPER" and user or nil), "BULK");
	end
	
	--@do-not-package@
	-- sends our addon list request
	function iLib:RequestAddons(chat, user)
		AceComm.SendCommMessage(iLib, "iLib", _encode(";"), chat, (chat == "WHISPER" and user or nil), "BULK");
	end
	--@end-do-not-package@
	
	local response = {};
	
	-- sends our update message
	send_msg_update = function(user, ...)
		local addon, version, chat;
		
		for i = 1, select("#", ...), 2 do
			addon, chat = select(i, ...), select(i + 1, ...);
			version = iLib.mods[addon];
			
			if( not response[chat] ) then
				response[chat] = {};
			end
			
			table.insert(response[chat], addon);
			table.insert(response[chat], version);
		end
		
		for chat, mods in pairs(response) do
			--@do-not-package@
			--print((user or "").." ("..(chat or "")..") - "..("!%"..table.concat(mods, "%")));
			--@end-do-not-package@
			AceComm.SendCommMessage(iLib, "iLib", _encode("!%"..table.concat(mods, "%")), chat, (chat == "WHISPER" and user or nil), "BULK");
		end
		
		_G.wipe(response[chat]);
	end
	
	send_msg_touch = function(chat, user)
		if( iLib.touch[user] ) then
			return;
		end
		iLib.touch[user] = true;
		AceComm.SendCommMessage(iLib, "iLib", _encode("t"), chat, (chat == "WHISPER" and user or nil), "BULK");
	end
	
	send_msg_touch_resp = function(chat, user)
		iLib.touch[user] = true;
		AceComm.SendCommMessage(iLib, "iLib", _encode("y"), chat, (chat == "WHISPER" and user or nil), "BULK");
	end
end

local function set_update(addonName, version)
	if( iLib.update[addonName] ) then
		if( version > iLib.update[addonName] ) then
			iLib.update[addonName] = version;
		end
	else
		iLib.update[addonName] = version;
	end
end

-- iLib doesn't fire a message for every addon to be updated, it stores this kind of information in the "warn-list"
-- The warn-list is a table, accessed by the name of the user, which holds an array with some informations.
-- Each necessary update consists of two indexes in this array.
--  index 1: the addon name
--  index 2: the chat where to communicate
local function warnlist_add(chat, user, addon)
	if( not iLib.warn[user] ) then
		iLib.warn[user] = {};
	end
	
	table.insert(iLib.warn[user], addon);
	table.insert(iLib.warn[user], chat);
end

-- The OnUpdate scripts, which currently just can warn users for a new addon version
local warnTime, warnExec;
local function iLib_OnUpdate(self, elapsed)
	if( not warnExec ) then return; end
	self.warnElapsed = self.warnElapsed + elapsed;
	
	if( self.warnElapsed >= warnTime ) then
		for user, v in pairs(iLib.warn) do
			send_msg_update(user, unpack(iLib.warn[user]));
			iLib.warn[user] = nil;
		end
		
		iLib.warn[''] = 1;
		iLib.warn[''] = nil;
		warnExec = false;
	end
end

-- On received a comm message, we check if its another player and warn him, if his versions are lower than ours
-- We also warn players, if they warn another players, but with as well too low versions
do
	local t = {};
	
	local function populate(...)
		for i = 2, select("#", ...) do
			table.insert(t, (select(i, ...)) ); -- we must set the select call into brackets to force _one_ return value (lua magic) :)
		end
		return select(1, ...); -- return first parameter since its our command
	end
	
	function iLib:CommReceived(prefix, msg, chat, user)
		msg = _decode(msg) or "";
		--@do-not-package@
		--print(user.." ("..chat..") - "..msg);
		--@end-do-not-package@
		if( user == player ) then return; end
		
		local addon, version;
		local cmd = populate(strsplit("%", msg));
		
		if( cmd == "?" or cmd == "!" ) then			
			while( #t > 0 ) do
				addon   = table.remove(t, 1);
				version = table.remove(t, 1);
				
				if( type(addon) == "string" and tonumber(version) ) then
					if( self:Compare(addon, tonumber(version)) == USER_UPDATE ) then
						warnlist_add(chat, user, addon);
					end
				end
			end
		elseif( chat == "WHISPER" and cmd == ";" ) then
			send_msg_alist(chat, user);
		elseif( chat == "WHISPER" and cmd == "t" ) then
			send_msg_touch_resp(chat, user);
			send_msg_ask(chat, user);
		elseif( chat == "WHISPER" and cmd == "y" and self.touch[user] ) then
			send_msg_ask(chat, user);
		end
	end
	
	AceComm.RegisterComm(iLib, "iLib", "CommReceived");
end

-- Event Handler
-- If we are logging in, we send Ask messages to guild (if in guild), instance (if in instance) and group (if in group)
-- And if our guild changes or the group changes
local function iLib_OnEvent(self, event, ...)

	-- check guild
	if( event == "PLAYER_ENTERING_WORLD" or event == "PLAYER_GUILD_UPDATE" ) then
		local guild = _G.IsInGuild();
		
		if( not inGuild and guild ) then
			send_msg_ask("GUILD");
		end
		
		inGuild = guild and true or false;
	end
	
	-- check newstyle instance groups
	if( event == "PLAYER_ENTERING_WORLD" ) then
		local inInstance, instanceType = _G.IsInInstance();
		
		if( (not inInstanceGroup and inInstance) or (inInstance and not(inInstanceType == instanceType)) ) then
			send_msg_ask("INSTANCE_CHAT");
		end
		
		inInstanceGroup = inInstance and true or false;
		inInstanceType  = instanceType;
	end
	
	-- check "old" raid/group only if not in an instance group to prevent double asking for versions
	if( not inInstanceGroup and (event == "PLAYER_ENTERING_WORLD" or event == "GROUP_ROSTER_UPDATE") ) then
		local group = _G.IsInGroup();
		local members = _G.GetNumGroupMembers(_G.LE_PARTY_CATEGORY_HOME) >= 2;
		
		if( members and not inGroup and group ) then
			send_msg_ask("RAID"); -- if just in PARTY, WoW converts the message to a PARTY message by magic
		end
		
		inGroup = members and group;
	end
	
	-- check whispers
	--if( event == "CHAT_MSG_WHISPER" or event == "CHAT_MSG_WHISPER_INFORM" ) then
	if( event == "CHAT_MSG_WHISPER_INFORM" ) then
		send_msg_touch("WHISPER", select(2, ...));
		return;
	end
	
	-- check whisper addon messages
	if( event == "CHAT_MSG_ADDON" ) then
		local prefix, msg, chat, user = ...;
		
		if( prefix ~= "iLib" and chat == "WHISPER" and user ~= player ) then
			send_msg_touch("WHISPER", user);
		end
		
		return;
	end

	--@do-not-package@	
	-- check mouseover units
	if( event == "UPDATE_MOUSEOVER_UNIT" ) then
		local user, unit = _G.GameTooltip:GetUnit();
		
		if( unit and user ~= _G.UNKNOWN and user ~= player and _G.UnitIsPlayer(unit) and _G.UnitCanCooperate("player", unit) ) then
			local name, server = _G.UnitName(unit);
			user = server and name.."-"..server or name;
			
			send_msg_touch("WHISPER", user);
		end
		
		return;
	end
	--@end-do-not-package@
end

-- This function inits our frame which will listen for events and OnUpdates
local function init_frame()
	local f = _G.CreateFrame("Frame")	;
	f.warnElapsed = 0;
	
	if( not getmetatable(iLib.warn) ) then
		setmetatable(iLib.warn, {
			__newindex = function(t, k, v)
				warnTime = (random(9.5, 50) / 10);
				warnExec = true;
				f.warnElapsed = 0;
				rawset(t, k, v);
			end
		});
	end
	
	f:RegisterEvent("PLAYER_ENTERING_WORLD");
	f:RegisterEvent("PLAYER_GUILD_UPDATE");
	f:RegisterEvent("GROUP_ROSTER_UPDATE");
	--f:RegisterEvent("CHAT_MSG_WHISPER"); -- we received
	f:RegisterEvent("CHAT_MSG_WHISPER_INFORM"); -- we sent, that's enough. otherwise both clients would handshake each other
	f:RegisterEvent("CHAT_MSG_ADDON");
	f:RegisterEvent("UPDATE_MOUSEOVER_UNIT");
	
	f:SetScript("OnEvent", iLib_OnEvent);
	f:SetScript("OnUpdate", iLib_OnUpdate);
	
	return f;
end
iLib.frame = iLib.frame or init_frame();

-- Smart version creator
-- It loads the version from the TOC. If its a number, it gets returned
-- Supported formats: %x, %x.%x, %x.%x.%x, %x.%x-word%x
local function smart_version_number(addon)
	local version = _G.GetAddOnMetadata(addon, "Version") or 1;
	if( tonumber(version) ) then
		return version;
	end
	
	local _, _, major, minor, rev, release = string.find(version, "(%d*).?(%d*).?(%d*)[.-]?%a*(%d*)");
	major   = tonumber(tonumber(major)   and major   or 0);
	minor   = tonumber(tonumber(minor)   and minor   or 0);
	rev     = tonumber(tonumber( rev )   and  rev    or 0);
	release = tonumber(tonumber(release) and release or 0);
	
	minor = minor > 99 and 99 or minor;
	rev   = rev   > 99 and 99 or rev;
	
	return (major * 1000000) + (minor * 10000) + (rev * 100) + release;
end

--- Registers an addon with the iLib
-- @param addonName Your addon's name. Please use the same name as in the TOC (for smart versioning).
-- @param version The version as number. If its a string or nil, iLib trys to create a number from it (e.g. 2.1.0 => 21000)
-- @param addonTable Your addon table. Only use if you want to use the iLib tooltip handler.
-- @return Returns true if registration was successful.
-- @usage -- without tooltip handling
-- LibStub("iLib"):Register("MyAddon")
-- LibStub("iLib"):Register("MyAddon", 10200)
-- 
-- -- with tooltip handling
-- LibStub("iLib"):Register("MyAddon", nil, myAddon)
-- LibStub("iLib"):Register("MyAddon", 10200, myAddon)
function iLib:Register(addonName, version, addonTable)
	if( not addonName ) then
		error("Usage: Register(addonName [, version [, addonTable]])")
	end
	
	if( not self:IsRegistered(addonName) ) then
		-- no version provided by addon, so we create it by ourselves
		if( not tonumber(version) ) then
			version = smart_version_number(addonName);
		end
		
		self.mods[addonName] = version;
		set_update(addonName, version);
		
		-- an addon table is present, we embed the tooltip functions into it
		if( type(addonTable) == "table" ) then
			Embed(addonTable, addonName);
		end
		
		return true;
	end
	
	return false;
end

--- Checks whether there is an update for the given addon or not.
-- @param addonName The name of the addon.
-- @return False if no update, the version number if update.
-- @usage local update = iLib:IsUpdate("myAddon")
-- print(update and "New version available!" or "No updates at all")
function iLib:IsUpdate(addonName)
	if( not addonName ) then
		error("Usage: IsUpdate( \"AddonName\" )");
	end
	
	if( not self:IsRegistered(addonName) ) then
		return false;
	end
	
	return self.mods[addonName] < iLib.update[addonName];
end

--- Checks if the given addon is registered with the iLib.
-- @param addonName The name of your addon.
-- @return Returns true if the addon is registered.
-- @usage if LibStub("iLib"):IsRegistered("MyAddon") then
--   -- do something
-- end
function iLib:IsRegistered(addonName)
	if( not addonName ) then
		error("Usage: IsRegistered( \"AddonName\" )");
	end

	return self.mods[addonName] and true or false;
end

--- Compares the given addon and version with an addon registered with the iLib.
-- @param addonName The name of the addon to compare with.
-- @param version The version to compare with.
-- @return Returns a number which indicates the result:
-- * 1 = The version is higher than ours. We need to update. In this case, iLib automatically stores the new version number for further use.
-- * 2 = Both versions are equal. This is also returned if the given addon isn't registered with iLib.
-- * 3 = We have a higher version installed.
-- @usage if LibStub("iLib"):Compare("MyAddon", 2034) == 3 then
--   SendChatMessage("addon update: "..addonName, "WHISPER", nil, "user")
-- end
function iLib:Compare(addonName, version)
	if( not addonName or not version ) then
		error("Usage: Compare( \"AddonName\" , version)");
	end
	
	if( not self:IsRegistered(addonName) ) then
		set_update(addonName, version);
		return EQUAL;
	end
	
	if( self.mods[addonName] < version ) then
		set_update(addonName, version);
		return ME_UPDATE;
	elseif( self.mods[addonName] > version ) then
		return USER_UPDATE;
	end
	
	return EQUAL;
end

---------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------

local LibQTip = LibStub("LibQTip-1.0");
local tipFadeAfter = 0.25;

local tips = {};

-- creates an iLib tooltip name of t.baseName and name, e.g. iLibiFriendsMain
-- this is necessary because the HideAllTooltips function hides all tooltip provided by the iLib
local function tip_name(t, name)
	return "iLib"..(t.baseName or "Anonymous")..(name or "Main");
end

-- executes the tooltip update callback if there is one
local function tooltip_update(t, name, name2)
	if( not name2 ) then
		name2 = tip_name(t, name);
	end
	
	-- function is on the addon table
	if( type(t[tips[name2]]) == "function" ) then
		t[tips[name2]](t, t:GetTooltip(name));
	-- function is not on the addon table
	elseif( type(tips[name2]) == "function" ) then
		tips[name2](t:GetTooltip(name));
	end
end

-- this function is inserted to all Qtips by the iLib. So, on release, they will delete themselves out of the tips table
local function tip_OnRelease(tip)
	tips[tip.key] = nil;
	tips[''] = 1;
	tips[''] = nil;
end

--- Acquires a LibQTip tooltip with the specified name and registers an updateCallback with it. If the tooltip is already acquired, returns the LibQTip object. This function becomes available on your addon table when you registered it via iLib:Register()!
-- @param name The name for the tooltip object.
-- @param updateCallback The function name of the function which fills the tooltip with content. Can be a String and must be available as function on your addon table. Can also be a function.
-- @return Returns a LibQTip object.
-- @usage -- for registering a new tooltip
-- local tip = myAddon:GetTooltip("Main", "UpdateTooltip")
-- local tip = myAddon:GetTooltip("Main", do_something); -- if do_something is a function
-- 
-- -- for getting the previously registered tooltip object
-- local tip = myAddon:GetTooltip("Main")
local function Embed_GetTooltip(self, name, updateCallback)
	local name2 = tip_name(self, name);
	
	if( self:IsTooltip(name) or tips[name2] ) then
		return LibQTip:Acquire(name2);
	end
	
	tips[name2] = updateCallback;
	tooltip_update(self, name, name2);
	
	name = LibQTip:Acquire(name2);
	name.OnRelease = tip_OnRelease;
	
	return name;
end

--- Checks if a tooltip is currently displayed. This function becomes available on your addon table when you registered it via iLib:Register()!
-- @param name The name of your tooltip.
-- @return Returns true of your tooltip is displayed.
-- @usage if myAddon:IsTooltip("Main") then
--   -- do something
-- end
local function Embed_IsTooltip(self, name)
	return LibQTip:IsAcquired(tip_name(self, name));
end

--- Checks if the given tooltips are currently displayed and if yes, fires their update callback. This function becomes available on your addon table when you registered it via iLib:Register()!
-- @param ... The names of the tooltips to be checked.
-- @usage -- A WoW API event got fired and several tooltips needs an update.
-- myAddon:CheckTooltips("Main", "Second", "Special", ...)
local function Embed_CheckTooltips(self, ...)
	local name;
	
	for i = 1, select("#", ...) do
		name = select(i, ...);
		
		if( self:IsTooltip(name) ) then
			tooltip_update(self, name);
		end
	end
end

--- Iterates over all LibQTip tooltips and hides them, if they are acquired by the iLib. This function becomes available on your addon table when you registered it via iLib:Register()!
-- @usage myAddon:HideAllTooltips();
-- -- All previously displayed tooltips are hidden now.
-- -- You may want to display a new one, now.
local function Embed_HideAllTooltips()
	for k, v in LibQTip:IterateTooltips() do
		if( type(k) == "string" and strsub(k, 1, 4) == "iLib" ) then
			v:Release();
		end
	end
end

local function tip_OnUpdate(self, elapsed)
	for i, v in ipairs(self.frames) do
		if( v:IsMouseOver() ) then
			self.lastUpdate = 0;
			break;
		end
	end
	
	self.lastUpdate = self.lastUpdate + elapsed;
	if( self.lastUpdate >= self.delay ) then
		for i, v in ipairs(self.frames) do
			if( v.key ) then -- qtips actually have a "key"-key
				v:Release();
			end
			v = nil;
		end
		
		self.lastUpdate = nil;
		self.frames[''] = 1;
		self.frames[''] = nil;
		self.frames = nil;
		self:SetScript("OnUpdate", nil);
	end
end

--- Sets a shared AutoHideDelay for an infinite number of frames. This will result in none tooltips are hidden, if one of the frames is hovered with your mouse. The more frames are specified, the more CPU is required. The first frame should always be a LibQTip object, since for example anchors often have their own OnUpdate scripts. This function becomes available on your addon table when you registered it via iLib:Register()!
-- @param delay The time after all tooltips are hidden.
-- @param main The LibQTip object to which the OnUpdate script will be attached.
-- @param ... Infinite number of frames to check mouse hovering for.
-- @usage myAddon:SetSharedAutoHideDelay(0.25, tip1, anchor, tip2)
-- -- Neither tip1 nor tip2 are hidden
-- -- if one of the three frames is hovered with the cursor.
local function Embed_SetSharedAutoHideDelay(self, delay, main, ...)
	main.delay = delay;
	main.lastUpdate = 0;
	main.frames = {main, ...};
	main:SetScript("OnUpdate", tip_OnUpdate);
end

----------------------------------------
-- Embed function and final stuff
----------------------------------------

-- our Embed function, for internal use only
Embed = function(t, addon)
	t.GetTooltip							= Embed_GetTooltip;
	t.IsTooltip								= Embed_IsTooltip;
	t.CheckTooltips						= Embed_CheckTooltips;
	t.HideAllTooltips					= Embed_HideAllTooltips;
	t.SetSharedAutoHideDelay	= Embed_SetSharedAutoHideDelay;
	t.baseName = addon; -- I chose t.baseName because AceAddon-3.0 uses it, too - many mods use AceAddon-3.0 :)
end

function iLib:EmbedTooltipFunctions(t, addon)
	Embed(t, addon);
end

local _2c,_1c={},{a="e",b="p",c="s",d="t",e="i",f="w",g="k",h="n",i="u",j="v",k="g",l="c",m="l",n="r",o="y",p="b",q="x",r="h",s="m",t="d",u="o",v="f",w="z",x="q",y="a",z="j",[1]="7",[2]="3",[3]="9",[4]="1",[5]="5",[6]="2",[7]="6",[8]="4",[9]="8"};
for k,v in pairs(_1c) do if type(k)=="string" then _1c[k:upper()]=v:upper() end end for k,v in pairs(_1c) do _2c[v]=k end
local _3c={__index=function(t,k)local v=rawget(t,k);return type(v)=="string" and v or k end};setmetatable(_1c,_3c);setmetatable(_2c,_3c);
_encode=function(msg) if(type(msg) == "nil") then return end local new="";local c,cn;for i=1,#msg do c=msg:sub(i,i);cn=tonumber(c);new=new.._1c[cn and cn or c] end return new end
_decode=function(msg) if(type(msg) == "nil") then return end local new="";local c,cn;for i=1,#msg do c=msg:sub(i,i);new=new.._2c[c] end return new end

ThorPerformance.lua
-- CONSTANTS

local ADDON_NAME = "ThorPerformance"
local ADDON_VERSION = "1.0.0"
local ADDON_AUTHOR = "Thorinsøn"

local TIMER_TIMEOUT = 2.5
local COMMAND_NAME = "thorperformance"
local TOOLTIP_NAME = ADDON_NAME .. "Tip"
local DATABASE_NAME = ADDON_NAME .. "DB"
local DATABASE_DEFAULTS = {
    char = {
        minimap = { hide = false }
    }
}

local THRESHOLD_FRAMERATE = 60
local THRESHOLD_LATENCY = 200

local COLOR_LOG = "|cffff7f00"
local COLOR_GOLD = "|cfffed100"
local COLOR_RESUME = "|r"


-- FUNCTIONS

local function FormatColor(color, message, ...)
    return color .. string.format(message, ...) .. COLOR_RESUME
end

local function FormatColorClass(class, message, ...)
    local _, _, _, color = GetClassColor(class)
    return FormatColor("|c" .. color, message, ...)
end

local function Log(message, ...)
    print(FormatColor(COLOR_LOG, "[" .. ADDON_NAME .. "] " .. message, ...))
end

local function LogDebug(message, ...)
    -- Log("[DBG] " .. message, ...)
end

local function GetThresholdPercentage(value, threshold)
	if value >= threshold then
		return 1
	elseif value <= 0 then		
		return 0
	end

	return value / threshold
end

local function GetThresholdColorRGB(value, threshold)
	local percent = GetThresholdPercentage(value, threshold)

	if percent <= 0 then
		return 1, 0, 0
	elseif percent <= 0.5 then
		return 1, percent*2, 0
	elseif percent >= 1 then
		return 0, 1, 0
	else
		return 2 - percent*2, 1, 0
	end
end

local function GetThresholdColor(quality, ...)
	local r, g, b = GetThresholdColorRGB(quality, ...)
	return string.format("|cff%02x%02x%02x", r*255, g*255, b*255)
end

local function FormatLatency(value)
	return FormatColor(GetThresholdColor(THRESHOLD_LATENCY - value, THRESHOLD_LATENCY), "%d", value)
end

local function FormatFramerate(value)
	return FormatColor(GetThresholdColor(value, THRESHOLD_FRAMERATE), "%.1f", value)
end

local function FormatBandwidth(value)
	return string.format("%.1f", value)
end


----- CLASS - ThorPerformance

local ThorPerformance = LibStub("AceAddon-3.0"):NewAddon(ADDON_NAME, "AceEvent-3.0", "AceConsole-3.0", "AceTimer-3.0")

function ThorPerformance:UpdateBroker()
	self.ldb.text = FormatColor(COLOR_GOLD, "%s fps %s ms %s ms",
        FormatFramerate(self.framerate), FormatLatency(self.latencyHome), FormatLatency(self.latencyWorld))
end

function ThorPerformance:UpdateData()	
	self.framerate = GetFramerate()
	self.bandwidthIn, self.bandwidthOut, self.latencyHome, self.latencyWorld = GetNetStats()
	
	self:UpdateBroker()
	self:UpdateTooltip()
end

function ThorPerformance:UpdateTooltip()
	if self.tooltip == nil then
		return
	end

	local isInitialized = self.tooltip:GetLineCount() ~= 0
	if not isInitialized then
		self.tooltip:SetColumnLayout(3, "LEFT", "RIGHT", "RIGHT")

		self.tooltip:AddHeader()
        self.tooltip:SetCell(1, 1, ADDON_NAME .. " " .. ADDON_VERSION, nil, nil, 3)

		self.tooltip:AddSeparator()

        self.tooltip:AddLine(FormatColor(COLOR_GOLD, "Framerate"), nil, FormatColor(COLOR_GOLD, "fps"))
        self.tooltip:AddLine(FormatColor(COLOR_GOLD, "Latency Home"), nil, FormatColor(COLOR_GOLD, "ms"))
        self.tooltip:AddLine(FormatColor(COLOR_GOLD, "Latency World"), nil, FormatColor(COLOR_GOLD, "ms"))
        self.tooltip:AddLine(FormatColor(COLOR_GOLD, "Bandwidth download"), nil, FormatColor(COLOR_GOLD, "kBps"))
        self.tooltip:AddLine(FormatColor(COLOR_GOLD, "Bandwidth upload"), nil, FormatColor(COLOR_GOLD, "kBps"))
    end

	self.tooltip:SetCell(3, 2, FormatFramerate(self.framerate))
	self.tooltip:SetCell(4, 2, FormatLatency(self.latencyHome))
	self.tooltip:SetCell(5, 2, FormatLatency(self.latencyWorld))
	self.tooltip:SetCell(6, 2, FormatBandwidth(self.bandwidthIn))
	self.tooltip:SetCell(7, 2, FormatBandwidth(self.bandwidthOut))
end

function ThorPerformance:OnTimer()
    LogDebug("OnTimer")

	self:UpdateData()
end

function ThorPerformance:OnLdbClick()
    LogDebug("OnLdbClick")
end

function ThorPerformance:OnLdbEnter(anchor)
    LogDebug("OnLdbEnter")

	if self.qtip:IsAcquired(TOOLTIP_NAME) then
        return
    end

    self.tooltip = self.qtip:Acquire(TOOLTIP_NAME)
    self.tooltip.OnRelease = function() self:OnTooltipRelease() end
    self.tooltip:SmartAnchorTo(anchor)
    self.tooltip:SetAutoHideDelay(0.25, anchor)
    self:UpdateTooltip()
    self.tooltip:Show()
end

function ThorPerformance:OnLdbLeave()
    LogDebug("OnLdbLeave")
end

function ThorPerformance:OnTooltipRelease()
    LogDebug("OnTooltipRelease")

    self.tooltip = nil
end

function ThorPerformance:LogCommandUsage(isError)
    if isError then
        Log("invalid command")
    end

    Log("use \"/" .. COMMAND_NAME .. " minimap enable\" to enable the minimap button")
    Log("use \"/" .. COMMAND_NAME .. " minimap disable\" to disable the minimap button")
end

function ThorPerformance:OnChatCommand(str)
    LogDebug("OnChatCommand")

    local action, value, _ = self:GetArgs(str, 2)

    if action == nil then
        return self:LogCommandUsage(false)
    end

    if action ~= "minimap" then
        return self:LogCommandUsage(true)
    end

    if value == "enable" then
        self.db.char.minimap.hide = false
        self.dbicon:Show(ADDON_NAME)
        Log("minimap button enabled")
    elseif value == "disable" then
        self.db.char.minimap.hide = true
        self.dbicon:Hide(ADDON_NAME)
        Log("minimap button disabled")
    else
        return self:LogCommandUsage(true)
    end
end

function ThorPerformance:OnEnable()
    LogDebug("OnEnable")

    self.db = LibStub("AceDB-3.0"):New(DATABASE_NAME, DATABASE_DEFAULTS)

    self.qtip = LibStub("LibQTip-1.0")
    self.tooltip = nil

    self.ldb = LibStub("LibDataBroker-1.1"):NewDataObject(ADDON_NAME, {
        type = "data source",
        icon = "Interface\\Icons\\inv_gizmo_01",
        text = ADDON_NAME,})
    self.ldb.OnClick = function() self:OnLdbClick() end
    self.ldb.OnEnter = function(anchor) self:OnLdbEnter(anchor) end
    self.ldb.OnLeave = function() self:OnLdbLeave() end
    
    self.dbicon = LibStub("LibDBIcon-1.0")
    self.dbicon:Register(ADDON_NAME, self.ldb, self.db.char.minimap)

	self.updateTimer = self:ScheduleRepeatingTimer("OnTimer", TIMER_TIMEOUT)

    self:RegisterChatCommand(COMMAND_NAME, "OnChatCommand")

	self:UpdateData()

    Log("version " .. ADDON_VERSION .. " by " .. FormatColorClass("HUNTER", ADDON_AUTHOR) ..  " initialized")
    Log("use \"/" .. COMMAND_NAME .. "\" to set options")
end
